src/backend/cpu/reduce.hpp

```diff
@@ -11,10 +11,15 @@
 #include <ops.hpp>
 
 namespace cpu {
-template<af_op_t op, typename Ti, typename To>
+template <af_op_t op, typename Ti, typename To>
 Array<To> reduce(const Array<Ti> &in, const int dim, bool change_nan = false,
                  double nanval = 0);
 
-template<af_op_t op, typename Ti, typename To>
+template <af_op_t op, typename Ti, typename Tk, typename To>
+void reduce_by_key(Array<Tk> &keys_out, Array<To> &vals_out,
+                   const Array<Tk> &keys, const Array<Ti> &vals, const int dim,
+                   bool change_nan = false, double nanval = 0);
+
+template <af_op_t op, typename Ti, typename To>
 To reduce_all(const Array<Ti> &in, bool change_nan = false, double nanval = 0);
-}  // namespace cpu
+}
```

Comment:
I'd say add back `// namespace cpu` here, but the file is indeed too short for that (though long-term, if this file grows, it'd be nice to have that comment).